### PR TITLE
VPN-5146: Create MZStepProgressBar

### PR DESCRIPTION
**Pull request by MattLichtenstein**
## Description

- Create `MZStepProgressBar` and `MZStepProgressBarDelegate` nebula components
- Create UI Testing view in Settings > Get help > Developer options for testing currently unused UI and components

https://github.com/mozilla-mobile/mozilla-vpn-client/assets/15353801/df4f98ce-8cfa-4e18-9830-b907f7bbe5ad

## Reference

[VPN-5146: Create MZStepProgressBar Nebula component](https://mozilla-hub.atlassian.net/browse/VPN-5146)

### CODE REVIEW

File: src/apps/vpn/translations/strings.yaml [hunk 1]
```diff
@@ -543,6 +543,24 @@ onboarding:
...
+    comment: Label for onboarding progress bar step
+  progressBarMorePrivacy:
+    value: Privacy
+    comment: Label for onboarding progress bar step
+  progressBarAddDevices:
+    value: Devices
+    comment: Label for onboarding progress bar step
+  progressBarGetStarted:
+    value: Start
+    comment: Label for onboarding progress bar step
```
Comment (flodolo) [inline src/apps/vpn/translations/strings.yaml:557]:
```suggestion
    comment: Label for onboarding progress bar step. It's the final step in the progress, and indicates that the user can now start using the app.
```

This is kind of hard to understand without the video. Feel free to tweak my suggestion.

File: src/apps/vpn/translations/strings.yaml [hunk 2]
```diff
@@ -543,6 +543,24 @@ onboarding:
...
+  progressBarAccessibilityStepComplete:
+    value: "%1. Onboarding step %2 of %3. Select to go back to this step."
```
Comment (flodolo) [inline src/apps/vpn/translations/strings.yaml:559]: Is %1 replace by Data, Privacy, etc. ?

Could we do something like this? `Onboarding step “%1” (%2 of %3). Select to go back to this step.`

**Review by flodolo — APPROVED** [review]

File: nebula/ui/components/MZStepProgressBarDelegate.qml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */import QtQuick 2.5
+import QtQuick.Layouts 1.15
+
+import Mozilla.Shared 1.0
+
+import compat 0.1
+
+Column {
+    id: delegate
+    property string source
```
Comment (lesleyjanenorton) [inline nebula/ui/components/MZStepProgressBarDelegate.qml:12]: `iconSource` or `imgSource` maybe?

File: src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+                    model: ListModel {
+                        id: stepProgressModel
+
+                        //Hacky workaround since QML list models can't take in scripts (the MZI18n localization script) as property values
+                        //This circumvents us having to make a C++ class for the progress bar delegates and a model for the progress bar as a whole
+                        //(i.e. navigationbarmodel.h/cpp and navigationbarbutton.h/cpp)
+                        //reference: https://stackoverflow.com/a/33161093
+                        property bool completed: false
+                        property string firstString: MZI18n.OnboardingProgressBarDataUse
```
Comment (lesleyjanenorton) [inline src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml:53]: What about handling this the same way as we do for `VPNSegmentedNavigation` ? 
https://github.com/mozilla-mobile/mozilla-vpn-client/blob/308692c7ee9d3bda41d8d6e4cd9af9dadc83c659/nebula/ui/components/MZSegmentedToggle.qml#L146
And then...
https://github.com/mozilla-mobile/mozilla-vpn-client/blob/308692c7ee9d3bda41d8d6e4cd9af9dadc83c659/src/apps/vpn/ui/screens/home/ViewServers.qml#L75-L84

File: src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+                        onSecondStringChanged: {
+                            if(completed) setProperty(0, "labelText", firstString);
+                        }
+
+                        onThirdStringChanged: {
+                            if(completed) setProperty(0, "labelText", firstString);
+                        }
+
+                        onFourthStringChanged: {
```
Comment (lesleyjanenorton) [inline src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml:70]: Why listen for these string changes and reset `labelText`?

File: nebula/ui/components/MZStepProgressBar.qml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                    easing.type: Easing.InOutQuad
+                }
+            }
+        }
+    }
+
+    Row {
+        id: delegateLayout
+
+        anchors.left: parent.left
```
[the remark this answers is not in the document]
Comment (MattLichtenstein) [reply nebula/ui/components/MZStepProgressBar.qml:49]: It being a generic nebula components make me want to say that we shouldn't enforce a maximum width, and we should let the user (instantiator) decide. I've updated the example in the UI testing dev settings, let me know what you think.

I can't think of a reason why it wasn't done this way for the navbar as well

File: src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml (hunk 4 above)
Comment (MattLichtenstein) [reply src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml:52]: That...is a good idea. Thanks

File: src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml (hunk 5 above)
Comment (MattLichtenstein) [reply src/apps/vpn/ui/screens/getHelp/developerMenu/ViewUiTesting.qml:69]: Obsolete after the above comment 🙂

**Review by lesleyjanenorton — APPROVED** [review]
Thanks @MattLichtenstein

File: nebula/ui/components/MZStepProgressBar.qml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            delegate:  MZStepProgressBarDelegate {
+                id: progressBarDelegate
+                source: model.source
+                labelText: model.labelText
+                labelWidth: progressBar.width / progressBar.model.count
+                currentState: {
+                    if (index === activeIndex) {
+                        MZStepProgressBarDelegate.State.Active
+                    }
+                    else if (index < activeIndex) {
```
[the remark this answers is not in the document]
Comment (lesleyjanenorton) [reply nebula/ui/components/MZStepProgressBar.qml:65]: I think it's more of a question for UX... if you've completed all steps, should navigating backwards through the steps change the state of the steps with an index greater than the `activeIndex`? 

For instance: 
1. you complete all the steps
2. you click on step 2
3. steps 3 and 4 are now State.Incomplete, even though you've actually "completed" them
4. you are now required to click the 'next' button to revisit step 3 or 4

[the remark this answers is not in the document]
Comment (MattLichtenstein) [reply nebula/ui/components/MZStepProgressBar.qml:65]: Yep, thats intentional. The affordance is more of a "Where are we currently at" in the flow vs "What have we completed" in the flow. I think navigating through a "full" (all steps completed) progress bar would be a weird experience.

I also had @peterbenvenuto confirm.

**Pull request merged**